Sources/AblyChat/Message.swift

```diff
@@ -3,16 +3,17 @@ import Foundation
 public typealias MessageHeaders = Headers
 public typealias MessageMetadata = Metadata
 
-public struct Message: Sendable {
+public struct Message: Sendable, Codable, Hashable {
+    
     public var timeserial: String
-    public var clientID: String
+    public var clientID: String?
```

Comment:
Please can you make sure the commit message explains these public API changes?

Reply:
clientID can be optional in the cocoa-sdk so when using Realtime to initialise this SDK, we can also have a null clientID... I'll add this to the PR description too

Reply:
`RealtimeClient#clientId` is nullable in all of our SDKs (see spec); why does it affect us but not JS?

Reply:
Looking at the `RealtimeClient` interface in ably-js seems to show clientId is not optional at all here. I'm guessing that's where our difference in implementation comes from. 

ably-js not adhering to the spec is another matter of course, but I'll run this by @AndyTWF to confirm I'm looking in the right places and this was by design in chat-js due to how ably-js works. 


<img width="754" alt="Screenshot 2024-10-08 at 21 03 42" src="https://github.com/user-attachments/assets/dee752de-8c7d-424b-8949-e81a5ba39998">

Reply:
I can't comment on the core SDKs - I wasn't aware of the nullablity of client id when we did chat-js.

One thing to mention though, is that Chat channels currently require a non-null, non-wildcard client id to be used - regardless :)

Reply:
clientId is `null` when connection is not identified, i.e. when user is not logged in.
When connection is `anonymous`, `clientId` is null. 
Though, when initializing client it's not recommended to set `clienId` as null, it is set by `token` itself. 
Also, there was recent PR to update the spec -> https://github.com/ably/specification/pull/195

Reply:
For more details read ->
https://sdk.ably.com/builds/ably/specification/main/features/#RSA12
https://sdk.ably.com/builds/ably/specification/main/features/#RSA7b
https://sdk.ably.com/builds/ably/specification/main/features/#RSA7c

Reply:
@AndyTWF This also means connection always needs to be identified when passing shared `realtime` instance with `chat wrapper sdk`

Reply:
Now we have external dependency to have non-null, non-wildcard clientId set on passed realtime instance. This will be dependent on their token auth mechanism used. Please note that, user can choose to keep `clientId` as null ( when user is anonymous, not logged in or guest user). 
We use this terminology in laravel-broadcaster, and we have null clientId in such a case.

Reply:
@umair-ably as per RSA7c, provided `clientId` can be null. Just that, it doesn't need to be a wildcard.

Reply:
> @AndyTWF This also means connection always needs to be identified when passing shared `realtime` instance with `chat wrapper sdk`, which might not always be the case.

Yep - this is a requirement of Chat. A number of chat features do require that you "know" who someone is, even if the clientID is randomly generated. Of course, if you want to do anonymous activity for other realtime use-cases, you can spin up a second realtime client.

Reply:
So, do we throw error if someone passing realtime instance without explicit clientId set on it ?

Reply:
Also, clientId is supposed to be unique or no ?

Reply:
@AndyTWF Given ably-js seems to be non-null to begin with, chat-js wouldn't have needed to handle this problem in the same way we do. 

Given ably-cocoa can instantiate Realtime with no clientId, but we also have this non-null requirement in Chat, would a random UUID as the clientId be sufficient for this requirement?

Reply:
Or I could throw an error if clientId is null, pushing the user towards initialising Realtime properly?

Reply:
@umair-ably your assumption of ably-js with non-null clientId is wrong. It can be set to null via token containing null clientId.

Reply:
Also, if you don't set clientId via clientOptions, it will be set to `null`. 
It is set to wildcard, only when basic auth is used via `API_KEY`, which is not recommended for client side applications. 
We can consider token auth for our use-case, which means it can be set as `null` via `token` payload.

Reply:
I think this is a complicated scenario ( based on my experience working with laravel ).
Involves hourly refreshing token, mutates auth clientId based on auth token success / failure. 
I believe we should have a proper spec to handle the same.

Reply:
My thoughts on this:

- given that, as Andy said, Chat has to be used with an identified client, we should _not_ change the public API of the Chat SDK to make `Message.clientId` nullable
- instead, we should figure out how the Chat SDK should behave in the case where `realtime.clientId` is for some reason `null`; I suggest that for now we treat it as a precondition failure and create an issue + TODO to return to this

Reply:
>  I suggest that for now we treat it as a precondition failure and create an issue + TODO to return to this

I agree with that too - I'll make that change

Reply:
I would also like to point out some breaking cases that will arise due to shared realtime client instance 
- Currently upgrading connection from unidentified to identified ( from null clientId to non-null clientId ) will break existing connection and get it into failed state.
-  Please take a look at detailed discussion here -> 
https://github.com/ably/ideas/issues/182. This is a very old issue that needs to be fixed.
- With connection going into failed state, all channels will get into failed state once clientId is mutated. We had to explicitly fix it in -> https://github.com/ably-forks/laravel-echo/pull/38/files

Reply:
`clientId` is non-optional again